code/modules/power/singularity/singularity.dm

```diff
@@ -446,3 +446,136 @@
+	var/list/inputs = list("up", "down", "left", "right")
+	/// deadchat_listener datum for the singularity.
+	var/datum/deadchat_listener/singularity_listener/listener
+
+/obj/singularity/deadchat_controlled/Initialize(mapload, starting_energy = 50)
+	..()
+	listener = new /datum/deadchat_listener/singularity_listener
+	listener.parent = src
+	GLOB.deadchat_listeners += listener
+	
+/obj/singularity/deadchat_controlled/Destroy()
+	..()
+	var/message = "<span class='deadsay italics bold'>The deadchat controlled singularity has been destroyed.</span>"
+	deadchat_broadcast(message)
+	GLOB.deadchat_listeners -= listener
+	qdel(listener)
+	
+/obj/singularity/deadchat_controlled/proc/process_deadchat(ckey, message)
+	if(deadchat_mode == "Anarchy")
+		var/cooldown = ckey_to_cooldown[ckey]
+		if(!cooldown)
+			ckey_to_cooldown[ckey] = NONE
+			cooldown = 0
+		if(cooldown > 0)
+			return
+		var/direction
+		message = lowertext(message)
+		switch(message)
+			if("up")
+				direction = NORTH
+			if("down")
+				direction = SOUTH
+			if("left")
+				direction = WEST
+			if("right")
+				direction = EAST
+		if(direction)
+			forceMove(get_step(src,direction))
+			eat()
+			ckey_to_cooldown[ckey] = TRUE
+			addtimer(VARSET_LIST_CALLBACK(ckey_to_cooldown, ckey, FALSE), anarchy_cooldown)
+	else if(deadchat_mode == "Democracy")
+		var/vote = ckey_to_cooldown[ckey]
+		if(!vote)
+			ckey_to_cooldown[ckey] = FALSE
+			vote = NONE
+		message = lowertext(message)
+		if(inputs.Find(message))
+			ckey_to_cooldown[ckey] = message
+
+/obj/singularity/deadchat_controlled/proc/democracy_loop()
+	if(QDELETED(src) || deadchat_mode != "Democracy")
+		return 
+	var/result = count_democracy_votes()
+	if(result != NONE)
+		forceMove(get_step(src, result))
+		eat()
+		var/direction_name
+		switch(result)
+			if(NORTH)
+				direction_name = "up"
+			if(SOUTH)
+				direction_name = "down"
+			if(WEST)
+				direction_name = "left"
+			if(EAST)
+				direction_name = "right"
+		var/message = "<span class='deadsay italics bold'>The singularity has moved [direction_name]!<br>New vote started. It will end in [democracy_cooldown/10] seconds.</span>"
+		deadchat_broadcast(message)
+	else
+		var/message = "<span class='deadsay italics bold'>No votes were cast this cycle. Type 'up', 'down', 'left' or 'right' to cast your vote!</span>"
+		deadchat_broadcast(message)
+	addtimer(CALLBACK(src, .proc/democracy_loop), democracy_cooldown)
+			
+/obj/singularity/deadchat_controlled/proc/count_democracy_votes()
+	var/list/votes = list(0,0,0,0)
+	var/found_vote = FALSE
+	for(var/vote in ckey_to_cooldown)
+		switch(ckey_to_cooldown[vote])
+			if("up")
+				votes[1]++
+			if("down")
+				votes[2]++
+			if("left")
+				votes[3]++
+			if("right")
+				votes[4]++
+		if(ckey_to_cooldown[vote] != NONE)
+			found_vote = TRUE
+		ckey_to_cooldown[vote] = NONE
+	if(!found_vote)
+		return NONE
+	if(votes[1] >= votes[2] && votes[1] >= votes[3] && votes[1] >= votes[4])
```

Comment:
These checks are cringe. Just return the key of the highest value

Reply:
**should** be fine now.